### DISCUSSION TITLE
Older dosemu2 doesn't work with shareable serial port IRQ (no EOI?), can this be detected?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1867] [Q&A]
The problem occurs for `Build 2.0pre8-20191102-1332-g9cf2b7560`

I recently updated lDebug to allow sharing an IRQ handler, eg for using COM2 and COM4 ports at the same time, which by default are set up to the same IRQ. One commit that explains this in part is https://hg.pushbx.org/ecm/ldebug/rev/01bcbb47b2fc

The difference is, when the DCO6 option 10000h is enabled (the default now), then the debugger will not issue an EOI and it will chain to the prior interrupt handler unconditionally, after it is done servicing its own interrupt reasons (if any). I want to not send EOI in my handler because I suspect that the In-Service Register wouldn't depict our interrupt as being in service any longer after the EOI has been issued.

The problem is as follows. When I set up COM2 with `socat pty,link=/tmp/vptty-dos,rawer pty,link=/tmp/vptty-linux,rawer & disown` and using `dosemu -dumb -td -kt -I "serial { com 2 device /tmp/vptty-dos }"` connecting to `picocom /tmp/vptty-linux` then if I run lDebug and enter `r dco or= 4000`, I get the keep prompt displayed as intended. However, after that display only one letter is accepted as input, any further input appears to go missing. Possibly relatedly, running `u ptr ri0Bp` (while not currently connected to a serial port) disassembles as a slide of HLT instructions. Clearing the DCO6 flag 10000h fixes this problem.

On a more recent version, `Build 2.0pre9-dev-20221127-489-g359c2468a`, everything works regardless of what the DCO6 flag is set as. Possibly related is that the default interrupt 0Bh handler issues EOI, as expected:

```
-u ptr ri0Bp
F000:F568 FA                cli
F000:F569 50                push    ax
F000:F56A B020              mov     al, 20
F000:F56C E620              out     20, al
F000:F56E 58                pop     ax
F000:F56F CF                iret
...
```

Can you explain why this doesn't work on the older dosemu2? Ideally I could find a way to automatically detect a dosemu2 not working with my IRQ sharing and disable it by default on those revisions.

**Comment by ecm-pushbx** [comment]
Program version for testing this: https://pushbx.org/ecm/download/old/ldebug/20221216.zip

**Comment by ecm-pushbx** [comment]
I added a detection for the problem of missing EOI. If the timer handler (not interrupt, just part of the program polled repeatedly by the `getc` function) finds that the In-Service Register indicates we're yet to send an EOI, outside our interrupt handler, then it will enable a new option and issue one EOI by itself.

The new option makes it so we always issue an EOI. If we also want to share the IRQ we call the next handler instead of chaining, then do the EOI after the next handler has returned. I hope that this is still compatible with a wide range of setups.

The main changeset is at https://hg.pushbx.org/ecm/ldebug/rev/187d9c5ed6b2

**Comment by stsp** [comment]
I think the fix was in 4617fbda59
Before that, there was no "default
handler". Why it was fixed so late?
Well, obviously because no one
reported that bug so I've found it
by doing the code review.
